src/AMSlib/AMSTensor.cpp

```
@@ -0,0 +1,215 @@
+    auto& rm = ams::ResourceManager::getInstance();
+    rm.deallocate(_data, _location);
+    _data = nullptr;
+    _owned = false;
+  }
+}
+
+AMSTensor::AMSTensor(AMSTensor&& other) noexcept
+    : _data(other._data),
+      _elements(other._elements),
+      _element_size(other._element_size),
+      _shape(std::move(other._shape)),
+      _strides(std::move(other._strides)),
+      _dType(other._dType),
+      _location(other._location),
+      _owned(other._owned)
+{
+  other._data = nullptr;
+  other._owned = false;
+}
+
+AMSTensor& AMSTensor::operator=(AMSTensor&& other) noexcept
+{
+  if (this != &other) {
+    // Free existing resources
+
+    // Steal resources from `other`
+    _data = other._data;
+    _elements = other._elements;
+    _element_size = other._element_size;
+    _shape = std::move(other._shape);
+    _strides = std::move(other._strides);
+    _dType = other._dType;
+    _location = other._location;
+    _owned = other._owned;
+
+    other._data = nullptr;
+    other._owned = false;
+  }
+  return *this;
+}
+
+AMSTensor AMSTensor::transpose(AMSTensor::IntDimType axis1,
+                               AMSTensor::IntDimType axis2) const
+{
+  // Ensure the axes are within bounds
+  if (axis1 >= _shape.size() || axis2 >= _shape.size()) {
+    throw std::out_of_range("Transpose axes are out of bounds");
+  }
+
+  // Create new shape and strides for the transposed tensor
+  auto newShape = _shape;
+  auto newStrides = _strides;
+
+  // Swap the specified axes in both shape and strides
+  std::swap(newShape[axis1], newShape[axis2]);
+  std::swap(newStrides[axis1], newStrides[axis2]);
+
+  // Create a new tensor with the same data, new shape, and strides
+  if (dType() == AMSDType::AMS_DOUBLE)
+    return view((double*)_data, newShape, newStrides, _location);
+  else if (dType() == AMSDType::AMS_SINGLE)
+    return view((float*)_data, newShape, newStrides, _location);
+
+  throw std::runtime_error("Unknow data type in transpose\n");
```

Do we want to throw an exception here or just return some error code (or an empty tensor)?

---

Will check the code and decide.